django/db/models/fields/composite.py

```diff
@@ -0,0 +1,146 @@
+            raise ValueError(
+                "The left-hand side of the 'in' lookup must be an instance of "
+                "MultiColSource"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(f"'{self.lhs.field.name}' must be a tuple or a list")
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must be a collection of tuples or lists"
+            )
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must have {len(self.lhs)} elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        exprs = []
+        cols = self.lhs.get_source_expressions()
+
+        for vals in self.rhs:
+            exprs.append(
+                WhereNode(
+                    [Exact(col, val) for col, val in zip(cols, vals)], connector=AND
+                )
+            )
+
+        return compiler.compile(WhereNode(exprs, connector=OR))
+
+
+class CompositeValue(tuple):
+    def is_set(self):
+        return all(value is not None for value in self)
+
+
+class CompositeAttribute:
+    def __init__(self, field):
+        self.field = field
+
+    def __get__(self, instance, cls=None):
+        return CompositeValue(
+            getattr(instance, field_name) for field_name in self.field.field_names
+        )
+
+    def __set__(self, instance, values):
+        if values is None:
+            values = (None,) * len(self.field.field_names)
+
+        for field_name, value in zip(self.field.field_names, values):
+            setattr(instance, field_name, value)
+
+
+class CompositePrimaryKey(Field):
+    descriptor_class = CompositeAttribute
+
+    def __init__(self, *args, **kwargs):
+        kwargs["db_column"] = None
+        kwargs["editable"] = False
+        super().__init__(**kwargs)
+        self.field_names = args
+        self.fields = None
+
+    def contribute_to_class(self, cls, name, **_):
+        super().contribute_to_class(cls, name, private_only=True)
+        cls._meta.pk = self
+        setattr(cls, self.attname, self.descriptor_class(self))
+
+    def get_attname_column(self):
+        return self.get_attname(), self.db_column
+
+    def __iter__(self):
+        return iter(self.fields)
+
+    def __len__(self):
+        return len(self.fields)
+
+    @cached_property
+    def cached_col(self):
+        return MultiColSource(self.model._meta.db_table, self.fields, self.fields, self)
+
+    def get_col(self, alias, output_field=None):
+        return self.cached_col
+
+    def get_lookup(self, lookup_name):
+        if lookup_name == "exact":
+            return TupleExact
+        elif lookup_name == "in":
+            return TupleIn
+
+        return super().get_lookup(lookup_name)
```

Comment:
I get a sense that all the other lookups won't work by default.

Reply:
If it's not implemented, it will throw a `django.db.utils.ProgrammingError` when executing the query - not sure if an explicit error is necessary.

Btw, I implemented `gte`, `gt`, `lte`, `lt` too now.

Reply:
This branching has been removed btw, I replaced it with a mixin as you suggested. We could throw an error if the tuple version is not implemented for a lookup.

Reply:
I refactored and the branch is back - will raise a `NotImplementedError` 👍